DietPi-Software | Box86 and Steam for armhf

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready
- [x] Get to work with Steam
- [x] Test Baba is You
- [x] Test Portal
- [x] Retest on a fresh installation

**Reference**: https://github.com/MichaIng/DietPi/issues/3957

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Software | Begin to add Box86
- DietPi-Software | Fix syntax
- DietPi-Software | Change software ID to avoid conflict with Portainer
- DietPi-Software | Fix Raspberry Pi detection
- DietPi-Software | Add uninstall
- DietPi-Software | Specify Raspberry Pi Models
- DietPi-Software | Add `make install`
- DietPi-Software | Add support for Steam on ARM
- DietPi-Software | Fix Steam uninstall
- DietPi-Software | Stop Steam from installing i386 packages on ARM
- DietPi-Software | Fixes
- DietPi-Software | Add warning about running x86 software on ARM
- DietPi-Software | Add Steam desktop shortcut
- DietPi-Software | Fixes
- DietPi-Software | Exclude steamdeps from being updated
- DietPi-Software | Remove steam desktop shortcut on uninstall

nonnorm committed:
Begin to add Box86

nonnorm committed:
CodeFactor Fixes

nonnorm committed:
Fix ==

nonnorm committed:
Again fix syntax

nonnorm committed:
Wrap everything in G_EXEC

nonnorm committed:
Remove extra ]

nonnorm committed:
Change software ID

nonnorm committed:
Remove gl4es install

nonnorm committed:
Fix Raspberry Pi detection

nonnorm committed:
Finish fixing Rpi detection

nonnorm committed:
Show output on makes

nonnorm committed:
Add uninstall

nonnorm committed:
Add more to uninstall

nonnorm committed:
Specify Rpi Models

nonnorm committed:
Add gl4es

nonnorm committed:
Remove gl4es

nonnorm committed:
Fix Rpi 4 detection

MichaIng commented:
Remember that we have compiled packages available: https://dietpi.com/meveric/pool/main/b/
Since those are build in a generic way to work on all armhf SBCs, it's at least work testing and comparing the results with the freshly compiled binaries.

They depend on gl4es, as this at least makes sense for all graphical software, but that dependency could be removed. We could consider to add this APT repository, even that it is meant for Odroids in the first place, on Box64 installs to allow easy updating.

nonnorm commented:
I can't really test it out, though, without the dependency being removed.  My only ARM SBC is a Raspberry Pi, and `libgl-odriod` isn't in the Raspbian package repository.

MichaIng commented:
```
curl -sSLf https://dietpi.com/meveric/meveric.asc | gpg --dearmor -o /etc/apt/trusted.gpg.d/meveric.gpg --yes
echo -e 'deb https://dietpi.com/meveric/ all testing\ndeb https://dietpi.com/meveric/ buster main' > /etc/apt/sources.list.d/meveric.list
G_AGUP
apt install box86-odroid
```
Indeed the `libgl-odroid` dependency could cause an issue for other armhf SBCs as it ships `libGL.so` (GL => GLES wrapper, basically gl4es, if I'm not mistaken) which might not work on other boards or conflict with other implementations for those.

nonnorm commented:
You can't easily install i386 software without the Debian repository.  This could be a problem, as manually downloading the package from archive.debian.org is quite a hassle, and so is resolving dependencies (I couldn't even get GNU Hello to work without `libc6`).

MichaIng commented:
Hmm, the Debian repository could be added of course, and `dpkg --add-architecture i386` to allow installing from it, but that could easily break the system if a armhf library is replaced with an i386 one. I'm not sure how it's supposed to be used or usually used, but probably not to run i386 software from the Debian repository but binaries from other sources instead?

nonnorm commented:
It is mainly used in Twister OS, and its main use cases there are running Steam (games) and Wine.

nonnorm committed:
Merge branch 'dev' into 'box86'

MichaIng commented:
Okay, WINE is an own topic anyway, as there is no Windows software available in Linux repositories. Steam is available as i386 package only: https://packages.debian.org/buster/steam
Probably it is possible to allow foreign i386 architecture for that single package only, to that 🤔.

Just found an official DEB package: https://steamcdn-a.akamaihd.net/client/installer/steam.deb (from: https://store.steampowered.com/about/)
It installs an APT repository: https://repo.steampowered.com/steam/
Ah, that is for self-updates only, nothing else in it 😄.
Okay, if applying a foreign arch for a single package only does not work, probably for a single repository.
Would be great if we could make it available via Box86 on ARM boards.

nonnorm commented:
I'm working with Steam (using the official package) right now.  So far, it updates but doesn't actually show the main window at all.  The good thing is, Box86 automatically detects it and runs it.  Not sure if this works for all x86 software.

nonnorm commented:
Fixed now, I think it was because I clicked the wrong button for the GPU split.  I got Steam to run and signed in successfully, though per https://github.com/ptitSeb/box86#a-note-about-steam I did have to install `libappindicator1`.  I am running the 2D indie game Baba is You, though it is quite choppy.

nonnorm committed:
Add make install to all options

nonnorm committed:
Merge branch 'box86' of https://github.com/ravenclaw900/DietPi into box86

nonnorm committed:
Add support for Steam on ARM

nonnorm committed:
Add dependencies for Steam games

nonnorm committed:
CodeFactor Fixes

nonnorm committed:
More syntax fixes

nonnorm committed:
Fix uninstall

nonnorm committed:
6.35
Remove steamdeps binary for ARM, as it takes lots of time to run and tries to install unneeded i386 binaries

nonnorm committed:
Fix syntax

nonnorm committed:
6.35
Fix more syntax

nonnorm committed:
6.35
Keep steamdeps binary currently

nonnorm commented:
Steam starts up fine, but it runs `steamdeps` to install i386 dependencies, which you can't get on Raspberry Pi OS.  It also takes a long time for Steam to start up when that runs.  You could modify the `steamdeps.txt` file which has the list of dependencies, but it gets recreated on updates.

nonnorm committed:
Stop Steam from installing i386 packages on ARM

nonnorm committed:
6.35
Again fix syntax

nonnorm committed:
6.35
Add missing $

nonnorm commented:
I can verify that it works with Steam on Raspberry Pi, could someone else test it on one of the ARMv7 ODROIDS (or ARMv7 something else)?  Also, choosing 'Remember Password' on Steam seems to cause LibGL errors and SegFaults, like reported in  https://github.com/ValveSoftware/steam-for-linux/issues/3415.

Joulinar commented:
Sorry I'm owning RPi boards only

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6314,6 +6335,41 @@ If you want to update ${aSOFTWARE_NAME[$software_id]}, please use its internal u
+			G_EXEC mkdir box86-0.1.8/build
+			G_EXEC cd box86-0.1.8/build
+
+			if [[ $G_HW_MODEL_NAME == *'RPi'* ]]
```
To reliably identify RPis and Odroids, use the following:
```
if (( $G_HW_MODEL < 10 ))
then
    ...RPi...
elif (( $G_HW_MODEL < 20 ))
then
    ...Odoid...
else
    ...
fi
```

nonnorm committed:
6.35
Fixes

nonnorm committed:
Merge branch 'dev' into box86

nonnorm committed:
6.35
Add warning about running x86 software on ARM

nonnorm committed:
6.35
Fix installation platforms on Steam

nonnorm committed:
6.35
Remove Steam APT repository since it doesn't update from that and it recreates steamdeps

nonnorm committed:
6.35
Modify desktop file to start without showing any errors and in small mode

nonnorm committed:
6.35
Add desktop shortcut instead of modifying it

nonnorm commented:
Okay, I think that this is the most user-friendly that we're going to get.
- There is a warning about possible performance issues on install.
- Running Steam from the desktop shortcut disables the warning about `libc6` not being installed, and it also automatically opens in small mode, the only mode that Steam 32-bit works in. 
 - It also doesn't try to install i386 dependencies at run using `steamdeps` (don't know why it is even needed, they could just ship the dependencies with the DEB package), which takes a long time and doesn't work anyway.  
- The APT repository is also removed, as I'm fairly sure the Steam app self-updates, and the repository just seemed to update the binary that installed said i386 dependencies.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1143,12 +1143,9 @@ DietPi-Software will decrypt and use it for software installs. You can change it
-		aSOFTWARE_AVAIL_G_HW_MODEL[$software_id,21]=1
+		# Native PC and ARMv7 only
+		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,1]=0
+		aSOFTWARE_AVAIL_G_HW_MODEL[$software_id,3]=0
```
Typo here I guess:
```
aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,3]=0
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -2438,6 +2449,17 @@ DietPi-Software will decrypt and use it for software installs. You can change it
+		if [[ ${aSOFTWARE_INSTALL_STATE[156]} == 1 && $G_HW_ARCH == 2 ]]; then
+
+			aSOFTWARE_INSTALL_STATE[$software_id]=1
+			G_WHIP_MSG 'Warning: The piece of software you are about to install is meant for the x86 platform.\n\nBox86 will be used to run it on ARM, however there may be performance and compatibally issues.'
```
Let's move this prompt to the interactive section around line 16,500 "# Install Info/Warnings" and keep the pre-requirements section fully non-interactive.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6056,21 +6078,21 @@ If you want to update ${aSOFTWARE_NAME[$software_id]}, please use its internal u
+				local nvidia=
+				dpkg-query -s nvidia-driver &> /dev/null && nvidia='nvidia-driver-libs-i386'
 
+				debconf-set-selections <<< 'steam steam/question select I AGREE'
```
This was required to make the `steam` package install from the Debian repository non-interactive. If the official one does not have such an interactive prompt, that debconf call can be removed.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6056,21 +6078,21 @@ If you want to update ${aSOFTWARE_NAME[$software_id]}, please use its internal u
+				dpkg-query -s nvidia-driver &> /dev/null && nvidia='nvidia-driver-libs-i386'
 
+				debconf-set-selections <<< 'steam steam/question select I AGREE'
+				G_AGI $nvidia
```
Since this may be an empty install (it does not fail, but parses the APT cache and checks for dependencies, even with no input 😄), let's merge it with the above lines:
```
dpkg-query -s nvidia-driver &> /dev/null && G_AGI nvidia-driver-libs-i386
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6368,6 +6390,41 @@ If you want to update ${aSOFTWARE_NAME[$software_id]}, please use its internal u
+				elif [[ $G_HW_MODEL == 4 ]]; then
+					G_EXEC cmake .. -DRPI4=1 -DNOGIT=1 -DCMAKE_BUILD_TYPE=RelWithDebInfo
+				fi
+				G_EXEC_OUTPUT=1 G_EXEC make -j4
```
Did you try it with some optimisation options and `strip`, to lower binary size (RAM usage) and speed up binary load time a bit? Also The number of core should match the number of available cores, instead of hardcoded 4. All currently supported RPi models have 4 cores anyway, but the same make and install command can then be used for all devices:
```
G_EXEC make CFLAGS='-g0 -O3' -j$(nproc)
G_EXEC strip --remove-section=.comment --remove-section=.note box84 # if "box84" is name and path of the binary
G_EXEC make install
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12370,6 +12427,26 @@ _EOF_
 			ln -s /mnt/dietpi_userdata/steam /root/.steam
+			if [[ $G_HW_ARCH == 2 ]]; then
+				# Stop Steam from trying to install i386 packages on ARM
+				G_EXEC rm /etc/apt/sources.list.d/steam.list 
```
The repository can not work anyway, without either amd64 or i386 architecture being added to Dpkg. The repo key can then be removed as well:
```
G_EXEC rm /etc/apt/{sources.list.d/steam.list,trusted.gpg.d/steam.gpg}
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12370,6 +12427,26 @@ _EOF_
+				# Stop Steam from trying to install i386 packages on ARM
+				G_EXEC rm /etc/apt/sources.list.d/steam.list 
+				cat << '_EOF_' > /usr/bin/steamdeps
+#!/bin/bash
```
It will execute a tiny little faster when using `dash` here 😉.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12370,6 +12427,26 @@ _EOF_
+				cat << '_EOF_' > /home/dietpi/Desktop/steam.desktop
+[Desktop Entry]
+Name=Steam
+Exec=sh -c "STEAMOS=1 STEAM_RUNTiME=1 /usr/bin/steam +open steam://open/minigameslist %U"
```
To avoid a continuously active shell process, we could use:
```
dash -c "STEAMOS=1 STEAM_RUNTiME=1 exec /usr/bin/steam +open steam://open/minigameslist %U"
```
or:
```
env STEAMOS=1 STEAM_RUNTiME=1 /usr/bin/steam +open steam://open/minigameslist %U
```
- `exec` replaces the wrapping shell with the called command, so the shell process will be gone.
- `env` is explicitly used to call commands with environment variables and itself does not stay as active parent process, so that one looks like a cleaner solution.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -14026,8 +14113,8 @@ _EOF_
 			Banner_Uninstalling
-			G_AGP steam
-			rm -Rf /{root,home/*}/.steam /mnt/dietpi_userdata/steam
+			G_AGP steam-launcher
```
Let's still purge the `steam` package for installs prior to v6.35 or in case it got installed otherwise.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -13262,6 +13339,16 @@ _EOF_
+			Banner_Uninstalling
+			rm /usr/local/bin/box86
+			rm /etc/binfmt.d/box86.conf
+			rm -R /usr/lib/i386-linux-gnu/
```
Could that be a bit dangerous, if there are any other i386 libraries installed? Although how and what for would that be when not for Box86. Not sure, or is there some "uninstall" make command where we can pull the files from that were actually installed together with Box86?

nonnorm replied:
There is, and it's basically what I used. `cmake -P cmake_uninstall.cmake -- -n` lists the files, including specific libraries in `/usr/lib/i386-linux-gnu/`.  I'll remove those instead of the whole directory.

MichaIng commented:
Great implementation. I don't know if steam updates itself, especially when running as non-root, that should actually not be possible, but the repository can anyway not be used on armhf since all it's packages are marked for amd64 or i386. And adding those as allowed foreign DPkg architecture potentially dangerous and adds a lot of overhead to APT calls. The contained "steam" package is only a dummy package to pull in the "steam-launcher" package. It would recover the obsolete steamdeps script, although this alone could be avoided with some `path-exclude` DPkg configuration.

I'll find out about foreign architectures for a single repo/package, but no big issue to skip it.

MichaIng commented:
Ah wait, interesting... The repository works without adding a foreign architecture. Not sure how, since he lists are clearly distributed in amd64 and i386 sub directories: https://repo.steampowered.com/steam/dists/stable/steam/
But the launcher package itself is for "all" architectures: https://repo.steampowered.com/steam/pool/steam/s/steam/

<details>
<summary>I just added list and key and would be able to install steam-launcher without and other change on my RPi:</summary>

```
2021-01-23 13:44:50 root@micha:~# apt install steam-launcher
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  adwaita-icon-theme bubblewrap dbus-user-session dconf-gsettings-backend dconf-service dictionaries-common emacsen-common file fontconfig glib-networking
  glib-networking-common glib-networking-services gsettings-desktop-schemas gtk-update-icon-cache hicolor-icon-theme hunspell-en-us iso-codes libaspell15
  libatk-bridge2.0-0 libatk1.0-0 libatk1.0-data libatspi2.0-0 libavahi-client3 libavahi-common-data libavahi-common3 libcairo-gobject2 libcairo2 libcolord2
  libcups2 libdatrie1 libdconf1 libdrm-amdgpu1 libdrm-common libdrm-nouveau2 libdrm-radeon1 libdrm2 libdw1 libegl-mesa0 libegl1 libenchant-2-2 libepoxy0
  libfribidi0 libgbm1 libgdk-pixbuf-2.0-0 libgdk-pixbuf-xlib-2.0-0 libgdk-pixbuf2.0-0 libgdk-pixbuf2.0-common libgl1 libgl1-mesa-dri libglapi-mesa libgles2
  libglib2.0-0 libglvnd0 libglx-mesa0 libglx0 libgraphite2-3 libgstreamer-gl1.0-0 libgstreamer-plugins-base1.0-0 libgstreamer1.0-0 libgtk-3-0 libgtk-3-common
  libgudev-1.0-0 libharfbuzz-icu0 libharfbuzz0b libhunspell-1.7-0 libhyphen0 libice6 libjavascriptcoregtk-4.0-18 libjson-glib-1.0-0 libjson-glib-1.0-common
  liblcms2-2 libllvm11 libnotify4 libnspr4 libnss3 libopenjp2-7 liborc-0.4-0 libpam-systemd libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0 libpixman-1-0
  libproxy1v5 librest-0.7-0 libsecret-1-0 libsecret-common libsensors-config libsensors5 libsm6 libsoup-gnome2.4-1 libsoup2.4-1 libtext-iconv-perl
  libthai-data libthai0 libunwind8 libvulkan1 libwayland-client0 libwayland-cursor0 libwayland-egl1 libwayland-server0 libwebkit2gtk-4.0-37 libwebpdemux2
  libwoff1 libwpe-1.0-1 libwpebackend-fdo-1.0-1 libx11-xcb1 libxaw7 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0 libxcb-render0 libxcb-shm0
  libxcb-sync1 libxcb-xfixes0 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxft2 libxi6 libxinerama1 libxkbcommon0 libxmu6 libxrandr2
  libxrender1 libxshmfence1 libxt6 libxxf86vm1 libz3-4 shared-mime-info x11-common xbitmaps xdg-dbus-proxy xkb-data xterm zenity zenity-common
Suggested packages:
  ispell | aspell | hunspell wordlist hunspell openoffice.org-hunspell | openoffice.org-core isoquery aspell colord cups-common libenchant-2-voikko
  libvisual-0.4-plugins gstreamer1.0-tools gvfs librsvg2-common liblcms2-utils lm-sensors xfonts-cyrillic
Recommended packages:
  librsvg2-common aspell-en | aspell-dictionary | aspell6a-dictionary at-spi2-core enchant-2 libgdk-pixbuf2.0-bin libglib2.0-data xdg-user-dirs
  gstreamer1.0-gl gstreamer1.0-plugins-base libgtk-3-bin notification-daemon mesa-vulkan-drivers | vulkan-icd gstreamer1.0-plugins-good
  gstreamer1.0-pulseaudio | gstreamer1.0-alsa gstreamer1.0-libav xdg-utils | steamos-base-files x11-utils
The following NEW packages will be installed:
  adwaita-icon-theme bubblewrap dbus-user-session dconf-gsettings-backend dconf-service dictionaries-common emacsen-common file fontconfig glib-networking
  glib-networking-common glib-networking-services gsettings-desktop-schemas gtk-update-icon-cache hicolor-icon-theme hunspell-en-us iso-codes libaspell15
  libatk-bridge2.0-0 libatk1.0-0 libatk1.0-data libatspi2.0-0 libavahi-client3 libavahi-common-data libavahi-common3 libcairo-gobject2 libcairo2 libcolord2
  libcups2 libdatrie1 libdconf1 libdrm-amdgpu1 libdrm-common libdrm-nouveau2 libdrm-radeon1 libdrm2 libdw1 libegl-mesa0 libegl1 libenchant-2-2 libepoxy0
  libfribidi0 libgbm1 libgdk-pixbuf-2.0-0 libgdk-pixbuf-xlib-2.0-0 libgdk-pixbuf2.0-0 libgdk-pixbuf2.0-common libgl1 libgl1-mesa-dri libglapi-mesa libgles2
  libglib2.0-0 libglvnd0 libglx-mesa0 libglx0 libgraphite2-3 libgstreamer-gl1.0-0 libgstreamer-plugins-base1.0-0 libgstreamer1.0-0 libgtk-3-0 libgtk-3-common
  libgudev-1.0-0 libharfbuzz-icu0 libharfbuzz0b libhunspell-1.7-0 libhyphen0 libice6 libjavascriptcoregtk-4.0-18 libjson-glib-1.0-0 libjson-glib-1.0-common
  liblcms2-2 libllvm11 libnotify4 libnspr4 libnss3 libopenjp2-7 liborc-0.4-0 libpam-systemd libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0 libpixman-1-0
  libproxy1v5 librest-0.7-0 libsecret-1-0 libsecret-common libsensors-config libsensors5 libsm6 libsoup-gnome2.4-1 libsoup2.4-1 libtext-iconv-perl
  libthai-data libthai0 libunwind8 libvulkan1 libwayland-client0 libwayland-cursor0 libwayland-egl1 libwayland-server0 libwebkit2gtk-4.0-37 libwebpdemux2
  libwoff1 libwpe-1.0-1 libwpebackend-fdo-1.0-1 libx11-xcb1 libxaw7 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0 libxcb-render0 libxcb-shm0
  libxcb-sync1 libxcb-xfixes0 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxft2 libxi6 libxinerama1 libxkbcommon0 libxmu6 libxrandr2
  libxrender1 libxshmfence1 libxt6 libxxf86vm1 libz3-4 shared-mime-info steam-launcher x11-common xbitmaps xdg-dbus-proxy xkb-data xterm zenity zenity-common
0 upgraded, 140 newly installed, 0 to remove and 1 not upgraded.
Need to get 93.9 MB of archives.
After this operation, 309 MB of additional disk space will be used.
Do you want to continue? [Y/n]
```
</details>

So we could keep the repository installed on armhf but replace the steamdeps script and exclude it for package upgrades, at best already before doing the package install itself:
```
echo -e 'path-exclude /usr/bin/steamdeps\npath-exclude /usr/lib/steam/bin_steamdeps.py' > /etc/dpkg/dpkg.cfg.d/98-dietpi-steam
echo -e '#!/bin/dash\nexit 0' > /usr/bin/steamdeps
```

nonnorm commented:
While this PR has been open, Box86 has updated to `0.2.0` from `0.1.8`.  Sadly, the GitHub releases API can't grab the latest tag, and you can't really parse JSON natively on the bash command line.

MichaIng commented:
We could parse the tags API page: https://api.github.com/repos/ptitSeb/box86/tags
```
local version=$(curl -sSfL https://api.github.com/repos/ptitSeb/box86/tags | mawk '/"name": /{print $4;exit}')
Download_Install "https://github.com/ptitSeb/box86/archive/${version:-v0.2.0}.tar.gz" # known v0.2.0 as fallback
```
The build dir doesn't need to match the version, does it?

nonnorm commented:
It does, as it expands into `box86-VERSION`, though without the `v` prefix.

MichaIng commented:
Ah of course, it's the "build" dir that is created while the parent dir is there already 👍. So then:
```
local version=$(curl -sSfL https://api.github.com/repos/ptitSeb/box86/tags | mawk '/"name": /{print $4;exit}')
[[ $version ]] || version='v0.2.0' # Fallback
Download_Install "https://github.com/ptitSeb/box86/archive/$version.tar.gz"
G_EXEC mkdir box86-${version#v}/build
G_EXEC cd box86-${version#v}/build
```

nonnorm committed:
6.35
Fixes and optimiztions

nonnorm committed:
CodeFactor Fixes

nonnorm committed:
6.35
More CodeFactor Fixes

nonnorm committed:
6.35
Exclude steamdeps from being updated

nonnorm committed:
6.35
Remove steam desktop shortcut on uninstall

nonnorm committed:
6.35
Fix syntax

nonnorm committed:
6.35
Fix warning

nonnorm committed:
6.35
Only show warning on ARMv7

nonnorm committed:
6.35
Use (()) instead of [[]]

nonnorm committed:
6.35
Fix removal of Steam desktop shortcut

MichaIng committed:
v6.35

+ DietPi-Software | Box86: Move to free software ID 62
+ DietPi-Software | Box86: Install cmake dependency parallel to source archive download
+ DietPi-Software | Box86: On uninstall, attempt to remove files only if they exist
+ DietPi-Software | Steam: Apply DPkg file exclusion prior to and install dependencies parallel to package download and install
+ DietPi-Software | Steam: Align desktop entry with other cases: Install to /usr/share/applications and link to users ~/Desktop dirs
+ DietPi-Software | Steam: Remove DPkg config on uninstall

MichaIng commented:
I moved it to software ID 62, which can now be safely used.

Only when an ID becomes free within the same DietPi version, it should not be reused directly, so that beta users do not get the install state reset on each RC update, where only the removed software install flash is aimed to be unset.

Looks great! I'll to last tests tomorrow.

MichaIng committed:
v6.35

+ DietPi-Software | Steam: It's still required to apply i386 arch to install the nvidia-driver-libs-i386 package. On Steam startup this is otherwise done anyway.

MichaIng committed:
v6.35

+ DietPi-Software | Steam: Mute dpkg-query output when checking for nvidia-driver package

MichaIng commented:
### Testing Steam
🈹 VM Stretch: Dependencies broken, since it tries to install `libgl1`, wich is only available on backports, but without explicit `-t stretch-backports` APT does not automatically pull dependencies from backports as well, so due to version mismatch it fails.
```
2021-01-24 13:52:02 root@VM-Stretch:~# sudo -u dietpi steam
Package libgl1-mesa-dri:i386 needs to be installed
Package libgl1:i386 needs to be installed
Package libc6:i386 needs to be installed
Package libgl1:amd64 needs to be installed
Package bubblewrap needs to be installed
Package xdg-desktop-portal needs to be installed
Package xdg-desktop-portal-gtk needs to be installed
Steam needs to install these additional packages:
    libgl1-mesa-dri:i386, libgl1:i386, libc6:i386, libgl1:amd64, bubblewrap, xdg-desktop-portal, xdg-desktop-portal-gtk
.......
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 libgl1 : Depends: libglx0 (= 1.1.0-1~bpo9+1) but it is not going to be installed
 libgl1:i386 : Depends: libglx0:i386 (= 1.1.0-1~bpo9+1) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
Press return to continue:
```
Two possible solutions:
1. 🈹 `apt install libgl1-mesa-glx libgl1-mesa-glx:i386` pulls the non-backports packages that "provides" libgl1
    - Does not work, the dependencies script still insists to install the actual libgl1 package.
2. 🈯 `apt install -t stretch-backports libgl1 libgl1:i386` pulls dependencies from backports as well

My goodness, this launcher is a total mess, it's APT checks take forever as it does multiple redundant checks on every single start. I am considering to remove this for x86 as well and install the dependencies manually within our script. It would also allow to start it as root user?

- [x] VM Buster
- [x] VM Bullseye
- [x] RPi Buster

### Testing Box64
- [x] RPi Stretch
- [x] RPi Buster
- [x] RPi Bullseye

nonnorm commented:
Steam doesn't usually let you start as root user, installing itself into `~/.local/share/Steam`.  However, the command line argument `-forceservice` should let you run it as root.  I don't know how it installs the APT dependencies.

MichaIng commented:
> However, the command line argument `-forceservice` should let you run it as root.

Thanks for the info. For testing on VM with VESA driver it is not trivial to allow non-root users init an X session, so only LighDM login would then allow starting steam from desktop for non-root users then. But depends on the graphics driver as well, in other cases it might work when user is added to "video" group.

> installing itself into ~/.local/share/Steam

Jep, the Debian package install to `~/.steam`, hence we created a symlink to `/mnt/dietpi_usedata/steam`. That would now need to be done for `~/.local/share/Steam`. But we made that for the root user only, while now steam won't start OOTB as root. Not sure how to handle best, keep an individual steam instance in each users home dir, skipping the symlink or the other only idea I have is to create a systemd service and/or own wrapper script which starts the service as own service user "steam" with home dir `/mnt/dietpi_usedata/steam` and the flags you already applied to the desktop entry.

And we need to get rid of that insane steamdeps script definitely. It takes minutes on every start, completely unacceptable. At least I got it now working on Stretch.

MichaIng commented:
`-forceservice` does not work (anymore). The check is hardcoded in the `steam` executable. It can be edited out while for testing it is nice, for end users it's right that they should not start a game as root user, actually. I'll verify that it's the same with the Debian package. So only good solution is then indeed a service user and a wrapper. On ARM we can then include box86 directly into this wrapper.
_________
Checked the Debian package, and it is sooo much cleaner, no double and triple wrappers that install dependencies, no checks, it starts immediately as root and non-root user, a simple single binary, nothing more, a much better thing to start wrapping a service around, actually. It is even quite easily possible to install the Debian Bullseye package with newest Steam version on Debian Stretch, since there are no dependencies missing.

nonnorm commented:
The Debian package does seem to be much better.  However, how would you install it on ARM?  While we could download it from the Debian package mirror, it and all of the packages that it requires are i386.

MichaIng commented:
I had a closer look, the Debian package is not thaaat much different, it's still a wrapper of wrappers, but has two differences that I like:
- `~/.steam` is used for the actual steam client (that is auto-downloaded and auto-updated as you said before) and all configs and installed games. So it is a bit less scattered and easier to move/link userdata to an external drive. The Valve package stores config to `~/.steam` but data/games to `~/.local/share/Steam`.
- It means that the behaviour for users that use our's or Debian Steam implementation already do not face lost data and different behaviour when doing a reinstall. Otherwise we would need to place a warning or migrate existing data structures etc... We don't have many users because it never worked really good (#1062) but I have it running fine here on VM, created a testing account etc.

Repacking the Debian Bullseye steam package for ARM has the benefit that we can do all required changes directly, e.g. the changed start menu/application entry (desktop icons for individual users should stay a manual step, as we might change the way when/how this is done), in case a box86 wrapper script (?) and such.

Btw, since I don't know yet how box86 works but your suggested desktop entry does not contain it, how did you start Steam via box86 on RPi so far?

nonnorm commented:
Like I said in https://github.com/MichaIng/DietPi/pull/3989#issuecomment-751413024, it seems to recognize Steam automatically and run it with the emulation.

MichaIng commented:
Ah I think that is what binfmt does. The same way I emulate ARM on x84_64 VMs: It's actually only a little advanced chroot (systemd-nspawn) but binfmt automatically calls the non-matching ARM binaries with qemu-static-arm(64): https://wikipedia.org/wiki/Binfmt_misc

Okay we could still do this directly so that binfmt would not be required for Steam anymore, and we'd be able to pass some box86 options, if we find some useful, but that's all optional then.

I can say that it works very reliable, so for users it's just downloading/installing the x86 binaries and execute them 🙂.

MichaIng commented:
Interesting, on a Raspbian Bullseye system I needed to install `python3-minimal` to compile Box86, else `cmake` fails with:
```
2021-01-26 16:41:37 root@Building-ARMv6-Bullseye:/tmp/DietPi-Software/box86-0.2.0/build# cmake .. -DRPI2=1 -DNOGIT=1 -DCMAKE_BUILD_TYPE=RelWithDebInfo
-- Could NOT find Python3 (missing: Python3_EXECUTABLE Interpreter)
CMake Error at CMakeLists.txt:40 (message):
  You need a Python interpretor, CMake will exit.
```
Not sure why, but on other test systems this was not required, probably Python 3 was installed already, I'll recheck.
**EDIT: Indeed, Python 3 is required for the build.**

MichaIng committed:
v6.35

+ DietPi-Software | Box86: Add python3-minimal as build dependency

MichaIng committed:
v6.35

+ DietPi-Software | Steam: Switch to own repack of the Debian i386 Steam package to allow install on armhf arch
+ DietPi-Software | Steam: Since Bullseye, nvidia-driver-libs-i386 has been removed but nvidia-driver-libs is multiarch, hence nvidia-driver-libs:i386 can be installed instead which matches the changed package suggestions. The same can be done on older distros as nvidia-driver-libs-i386 is a meta package to pull nvidia-driver-libs:i386 only and the preinst check, introduced with Buster, prints a related info based on nvidia-driver-libs:i386 install, not on nvidia-driver-libs-i386 install, if an Nvidia GPU is detected. As well the Nvidia driver detection is changed to match the one from the preinst script which checks the kernel module directly instead of depending on an actual DEB package install.

MichaIng committed:
v6.35

+ DietPi-Software | Box86: Reload binfmt on all ARM systems rather than RPi only.

nonnorm committed:
v6.35
Fix typo

MichaIng commented:
Btw, desktop entry change and merging the additional ARMv7 dependencies into the package is still outstanding, but I want to test this step by step.

MichaIng commented:
Okay ready for testing.
- I added BOX86_LD_LIBRARY_PATH and BOX86_PATH variables to the Debian Steam launcher, so that Box86 finds the i386 libraries and binaries, if not available natively. With this, there are first still errors shown about missing libraries, but checking Box86 verbose level 1 (I enabled it for now by default) shows that it invokes the emulated i386 libraries instead. I'm not sure whether a native wrapped armhf libraries works better than an emulated i386 one, but let's see how it works.
- STEAMOS=1 is as well added to the launcher, so it does not need to be done via desktop shortcut.
- The STEAM_RUNTIME=1 has no effect, as it is enabled by default, if not set.
- There is one error about missing zenity command, but that's a bug in the Steam internal second launcher, respectively setup.sh script, which is called and uses zenity before the framework binaries have been added to PATH. It's visual only, so nothing to worry about.

Probably also other modes are possible now that Box86 can invoke the libraries correctly. Ah I see https://github.com/ptitSeb/box86#notes-about-steam now, but let's test it at least. I tested only on an ARMv7 container in x86_64 VM, but calling the X server from within the guest does not fully work. I'll setup an RPi image now to test it natively. Damn thing that I have a few ARMv8 SBCs, but those cannot run Box86 😢.

Joulinar commented:
I have a RPi3B+ and 4B if needed

MichaIng commented:
Jep, the more the better, especially to test the different modes and a few more games, when we find time. Some require additional libraries and I'm not sure if all of them are either pulled as dependency by the Debian package already or included in the Steam framework.
______________
Stories that happen: I took the SD card out of the NanoPi R2S to flash for RPi testing. The slot mechanism has quite some power and shot the SD card out, as I was not too careful. Now there is currently an open suitcase filled with a mountain of clothes, directly where the SD card shoot into (we needed to get rid of a wardrobe and stored clothes into suitcase temporarily). A tiny black chip somewhere in the middle of all those clothes. Took a while until I carefully sorted each piece of clothes out to find it fallen down to the bottom of the suitcase... 😄.

Joulinar commented:
Does it require a Steam account?

MichaIng commented:
Ah yes, that is true, however it's for free, including some games. But true that we cannot test non-free games then, at least when not playing them anyway.

MichaIng commented:
Another question is if we should ship GL4ES as well, a GL-to-GLES wrapper for ARMs to run GUI applications that use OpenGL only. But we need to compile it and make a package at best to apply conflicts and replacement flags for OpenGL library packages: https://github.com/ptitSeb/gl4es
On RPi we require two packages at best, one for legacy GPU driver and one for the VC4 KMS driver, if I understand the compiling instructions correctly. This means with a VC4 overlay/resolution enabled, if I'm not mistaken, the following can be installed to test: https://dietpi.com/meveric/pool/main/libg/libgl-odroid/libgl-odroid_20201213-1.1.5-1_arm64.deb
It's GL4ES.

nonnorm commented:
I could (again) test out some non-free games, including a Source Engine one.

MichaIng commented:
I added now `+open steam://open/minigameslist` to the launcher and a hence simplified the desktop entry. I'm still not 100% sure if libappindicator1 is required or not, sometimes I run into a crash on Steam start on RPi but it may not be related to the library.
To avoid Box86 recompiling: https://dietpi.com/downloads/binaries/buster/steam_armv7l.deb

When testing games, note that due to desktop install RPi GPU memory is raised to 64 MiB, which might not be enough for many games. Not sure if we should raise it to 128 MiB by default?

nonnorm commented:
Before I installed `libappindicator1`, I couldn't even make it past entering my password (Steam froze up and needed to be `SIGKILL`ed).  Not sure about the game memory, but all the games I played did seem to be at least a little choppy.

MichaIng committed:
v6.35

+ DietPi-Software | Steam: Do not override existing desktop entry, but force re-creation of broken symlinks

Joulinar commented:
Guys,
I did a test on my RPi3B+. First I installed box86 and rebooted the system. Afterwards I selected Steam to be installed. Surprisingly box86 was installed again, incl compiling. Usually boix86 should have been detected as already installed, isn't it?

nonnorm commented:
I created this issue: https://github.com/MichaIng/DietPi/issues/4018 in response to that.  The DietPi-Software dependency system by default always reinstalls everything, regardless of whether it's been installed before or not.

MichaIng commented:
Generally it is currently required to do reinstalls of dependencies, since they may be configured to work with the dependencies and at least to assure that latest configuration (e.g. also the systemd unit, user, permissions etc) and latest versions are applied which dependencies may depend on. At least this general behaviour cannot be undone without deeper review and testing each case.

However, in this particular case, I don't see a need to force a reinstall. There has been a recent commit to Box86 that affects stability of Steam, probably more will come, but especially for such long taking compiling install, we can suggest ARMv7 Steam users during dietpi-update to `dietpi-software reinstall 62` manually, if we know about a significant improvement.

Done: https://github.com/MichaIng/DietPi/pull/3989/commits/c07c2e15816226d939d3e6bdeeb942c20405e568

MichaIng committed:
v6.35

+ DietPi-Software | Box86: Do not force a reinstall if pulled in as dependency, since compiling can take quite long and all dependency specific configurations are done at the dependency-side, e.g. environment variables for Steam applied within the Steam launcher.

Joulinar commented:
probably I did something wrong but this is what I end up running a default installation of Steam

![image](https://user-images.githubusercontent.com/47155374/106472513-09b16680-64a3-11eb-9167-60f2770f5466.png)

nonnorm commented:
You need to open it in small mode (should be done automatically now?).  `View -> Small Mode`

MichaIng commented:
Check the `/usr/games/steam` file, which contains the arguments to call it in small mode, but somehow it doesn't work for me either. On first install start the regular browser is shown (which cannot work on ARMv7) and on subsequent starts, it is only started in background (task bar icon appears). Not sure what is wrong, I also verified that those arguments are correct, according to some Steam forum threads.

Probably the other two arguments break it somehow? I'll compare with the Valve launcher where it worked.
**EDIT: Yes `-nominidumps -nobreakpad` are not present in the Valve launcher.**

Ah I think I got the issue:
- There is still `"$@"` as last argument in the launcher, which means the last argument is defined as empty.
- Either this needs to be removed, which means adding other cmd options is not possible,
- or `+open steam://open/minigameslist` needs to be moved behind `"$@"`, or
- We remove it there and add it to the desktop icon (as it was before I messed with things 😅), so that users are still able to launch it in background only or in "big pictures" mode, which seems to work as well, as I read?

Joulinar commented:
`big pictures` was quite annoying for me. My RPi3B+ was running on nearly 100% CPU, the Gui was extremely slow and my mouse was not working.

MichaIng commented:
Good to know, it works but doesn't make much sense. However, to stay flexible, I'll now remove the arguments from the launcher and add it to the desktop entry instead, replacing the `%U` so that it hopefully opens in minimode. So starting the launcher itself won't open anything, i.e. open as panel icon, as long as no login is required ("remember me").

MichaIng commented:
Interesting, I just recognised that an additional http://packages.microsoft.com/repos/code repository was added. This must be installed by the Steam client 🤔.

It contains three MS Visual Studio related packages. Shall we do something about it?

nonnorm commented:
I sometimes had trouble with "remember me", with it causing Steam to crash on startup.  I had to run `steam --reset` each time to fix it.

nonnorm commented:
I don't think that's actually related, it happened on my "main" system (actually caught it happening, `adding vscode repo` during an `apt upgrade`).  I also saw it in this issue: #4076.
**EDIT:** Same thing here too https://www.reddit.com/r/linuxquestions/comments/l7ecud/adding_vscode_repo/.

Joulinar commented:
> **EDIT: Same thing here too https://www.reddit.com/r/linuxquestions/comments/l7ecud/adding_vscode_repo/.**

This I had on 2 empty RPi installations today. On 32bit as well as 64bit

nonnorm commented:
Strangely, but hopefully meaningless, the only reports seem to be on DietPi systems.

MichaIng commented:
It's from the raspberrypi-sys-mods package. Check `/var/lib/dpkg/info/raspberrypi-sys-mods/postinst`. Strange since the sources do not contain this: https://github.com/RPi-Distro/raspberrypi-sys-mods/blob/master/debian/raspberrypi-sys-mods.postinst

This download contains it: http://archive.raspberrypi.org/debian/pool/main/r/raspberrypi-sys-mods/raspberrypi-sys-mods_20210125_armhf.deb

See: https://www.raspberrypi.org/forums/viewtopic.php?t=301011

And we cannot even block this effectively, since it is added via postinst script and it is even renewed, if made empty. We can comment the entry at least.

> the only reports seem to be on DietPi systems

Please don't say something like this before checking back the RPi (or manufacturer or Armbian) forum, I was shocked for a few minutes that something got compromised or so 😅.

Joulinar commented:
wow nice discussion on the RPi board. And @MichaIng is right in the middle 🤣

MichaIng commented:
Yeah, I mean this deserves some noise so that hopefully whoever was responsible learns that such change needs to be implemented a bit careful, with some info, or a debconf question, a file name prefix, anything so that we quickly know what it is. Without the info from you guys that it showed up during this packages install, I would have been completely lost: `dpkg -S /etc/apt/sources.list.d/vscode.list` was the only idea I had, giving no result, then reinstalled Steam package, removed Steam instance to have rerun the client installer, as this was my first guess, nothing, searching it on web shows that it does not (yet) show any hit in combination with Raspberry Pi, so how am I supposed to find out who or what put it there?

`grep -r 'vscode.list' /var` would have found it, but no idea how long I would have searched around until trying that...

Joulinar commented:
let's see if someone from our users will notice it. At least we already have the answer 🤣

MichaIng commented:
Back to topic:
- Indeed `libappindicator1` is strictly required.
- I was able to reproduce the failure when using "remember password". @Joulinar does this work in your case? Otherwise we should add it to the documentation, aside of the info that small mode works only.
- Big pictures is horrible indeed, it starts but not practically usable.

Joulinar commented:
for me Steam was crashing this afternoon. But I'm not sure at which stage. Atm Steam is not starting anymore. It's doing the update  check but that is it...

MichaIng commented:
Have you selected to remember the password?

I updated the package with the mentioned last changes. Works now as good as it get's, indeed not fantastic, but I have hope that things get a bit more stable by times with Box86 updates.
- https://dietpi.com/downloads/binaries/buster/steam_armv7l.deb

nonnorm commented:
If you run the `steam` command, you should get an error along the lines of:
```text
(box86:8299): GLib-GObject-WARNING : 16:35:29.056: ../../../gobject/gsignal.c:1656: signal "device_added" already exists in the 'NMObject' class ancestry

(box86:8299): GLib-GObject-WARNING : 16:35:29.056: ../../../gobject/gsignal.c:1656: signal "device_removed" already exists in the 'NMObject' class ancestry

(box86:8299): GLib-GObject-WARNING : 16:35:29.056: ../../../gobject/gsignal.c:1656: signal "permission_changed" already exists in the 'NMObject' class ancestry

(box86:8299): GLib-GObject-CRITICAL : 16:35:29.057: g_object_new_is_valid_property: object class 'NMClient' has no property named 'dbus-connection'
```

MichaIng commented:
I saw this by times, but doesn't seem to break start? Can be muted via `apt install libnm0`.
It's the NetworkManager library, but since we don't use it, it shouldn't have an effect.

nonnorm commented:
Whoops, I forgot to paste the end in.  It's quite possible that that is not related whatsoever, it's just the last thing before a `Segmentation Fault`.

MichaIng committed:
v6.35

+ DietPi-Software | Steam: Remove ARM dependencies from install code since it is now added to the package itself
+ DietPi-Software | Steam: Remove leftover symlinks on uninstall

MichaIng commented:
What I didn't check is if adding `libnm0` probably fixes remembered password? I was wondering since AFAIK it worked the last time I tested it where I'm not sure if I had that installed. But probably it's indeed unrelated. Strange is that those errors do not always show up.
___________
From my end it's ready to be merged. If `libnm0` does fix something, I can add it to the package dependencies.

Joulinar commented:
At least I'm able to start Steam client. Need to say, I never used it before. So not sure how to really start something 🙄 

![image](https://user-images.githubusercontent.com/47155374/106528635-5ae44900-64e9-11eb-91d9-ee88c5a24288.png)

MichaIng commented:
Yes, I also didn't find a way to install games from the small mode. So this only works when one uses Steam on a different client already and has games in it's libraries, then they show up in the mini games list as well to be installed locally. Or @ravenclaw900 do you know a way to search and install new games with only small mode?

nonnorm commented:
The best way I can think of is to get the games from https://store.steampowered.com instead.  They should then show up grayed-out in the menu, and you can double-click (or right click and select `Install Game`) to install them.

MichaIng commented:
Another good info for the docs 👍.

Joulinar commented:
I tried to add anything from the web site but somehow it did not show up on the client

EDIT: one more thing I noticed. If I close the client, from inside the menu, I'm not able to start the client again

MichaIng commented:
> EDIT: one more thing I noticed. If I close the client, from inside the menu, I'm not able to start the client again

Same here. It is not really stopping, but hanging (can be seen in `htop`). It needs to be killed (the steam.sh wrapper script), then it can be started again. When stopping it from the panel icon, it works.
______________________
Known issues:
- Only small mode supported
- Hence no possibility to search/add games from the client, which needs to be done from the website store or a different client
- When choosing to remember the login password, subsequent client starts fail
- Stopping the client via small mode menu does not really stop it

MichaIng commented:
Another issue:
- I added "Wizards" to my games list.
- On a fresh x86 install, I can double-click it to have it downloaded and started.
- On RPi I could do the same, until the window that informs me that it's now downloaded in the background. It then shows to be installed immediately, but there was nothing actually downloaded. Double-clicking the game in the list prompts an error that the executable is missing, and of course it is, nothing has been downloaded.

Also, as well on x86, the "Play Game" button at the website does not work when used on the same system where our Steam is installed. I reverted the desktop entry to take URLs, but that does not work. I'm not sure how the handler for `steam://` URLs is supposed to work or implemented for Linux clients. I'll try it with the official package as well, just to rule out an issue with either the Debian package or the fact that the Bullseye package is installed on Buster.

And how do I actually "add" a game from the website? I only see the "Play Game" button, which does not work, but no button to add it to the own list (whichlist does not work), at least not for free games.

Exiting Steam via menu btw works on x86, so that is an ARM-only issue, or probably it just takes very long? The panel icon should disappear once Steam has fully exited.

Joulinar commented:
for me these are the 2 main challenges. Not able to start/download games without having the client installed somewhere else. And the client not being able to close correctly.

MichaIng commented:
Autologin (remember password) btw works fine on x86 without `libnm0`, that that is not related. A previous issue of Steam on x86 where sound in games did not work btw also does not exist anymore. At least on my VM, sound plays fine. In some cases it might be necessary to enable the ALSA software conversion plugin (possible now via dietpi-config Audio Options), if the sound sources are not in a format that is natively supported by the sound driver/card.

> Not able to start/download games without having the client installed somewhere else.

Yes, currently it's the same for me. I'll see if that works with the official launcher.

MichaIng commented:
The official steam launcher package is really not great. Since it is not marked as i386 package, but as "all", it pull in all dependencies as amd64 packages on x86_64 systems, while i386 variants would be required. So the latter need to be installed manually (at least some, until the dedicated dependency strict can be launched to check and install more i386 packages, which takes very long and is done on every Steam start) while the prior are obsolete. Probably doubled obsolete packages could be avoided by installing the i386 ones manually first, but maintaining such a dependency list is a nightmare as well. We did the right decision to use the Debian package 😄.

nonnorm commented:
Oddly, installing `libnm0` did seem to fix the "Remember Password" problem.  Do you want to add it as a dependency?

MichaIng commented:
Okay, strange that this is not required on x86 🤔.

I just recognised that `libappindicator1` and `libindicator7` are not available on Debian Bullseye.

<s>Could you try if the following preserves Steam startup:
```
apt-mark manual libdbusmenu-gtk4
dpkg -P --force-depends libappindicator1 libindicator7
```
</s>

____________
Okay, also with official Steam client, the "Play Game" button from the store website does not work, so it is a general issue with the Linux client, at least when using Firefox browser.
On x86 of course it doesn't matter much since the games browser in the client itself can be used.
Ah, in the client it seems the exact same website is present, so those buttons which link to steam:// URLs seem to be actually made for the clients, not to be used from a regular web browser?
____________
Lol, on x86 with official launcher package as well as Debian package on x86, the GLib-GObject NMClient warnings/errors show up as well (without breaking anything). I installed `libnm0` to see what changes on startup logs, and: the errors remain, but are fancy coloured ("WARNING" become yellow) and a blue coloured public IP address was added to the log like in @ravenclaw900 you logs as well without libnm0. Although it is actually libglib2.0-0 (or one of it's dependencies) which enables the colours and IP in the logs. It's however strange that obviously different libraries are used or required for some features on ARM compared to x86.

nonnorm commented:
The `steam://` links work on my computer (using Firefox), so I think it's a limitation of Linux.

MichaIng commented:
- `libnm0` fixes autologin as well in my case. Added it to the package dependencies.
- `libappindicator1` is definitely required, not only a dependency of it. How to handle it on Bullseye is a future issue, probably the package will still be added, we'll see.
- On exit (via menu AND via panel icon) an illegal thread termination error message is thrown which hangs Steam, I couldn't find any fix for that.

We should report the issues and the autologin fix to Box86 so that it can be added to their readme.

I'll merge the pull request now, further debugging and fine tuning can be done at a later time.

nonnorm commented:
Just for the record, I reported the Steam issues at https://github.com/ptitSeb/box86/issues/326.
